### ISSUE TITLE
kapp controller pull the imgpkg offline

### ISSUE THREAD

**Issue by debankur1** [carvel-dev/kapp-controller#1269]
Can the Kapp controller pull the image and other artifacts from the local imgpkg tarball?

**Comment by 100mik** [comment]
This repository is for the CLI `kapp` and not the controller. I will move this issue to the appropriate repo.
However, `kapp-controller` will only pull from a registry. I am struggling a bit to imagine why we would want this as kapp-controller runs on the cluster.

Could you elaborate on your use case? I think we are missing some bits here.

**Comment by debankur1** [comment]
think about an Edge cluster rather  a single node cluster and it is Airgapped having no access to any registry 
we have the freedom to load the images  locally to the container runtime and we can also facilitate the imgpkg tar ball in the host itself

**Comment by debankur1** [comment]
Any update on this ?

**Comment by 100mik** [comment]
Ah I see, I believe right now the only way forward here would be to spin up a local registry on that node.

**Comment by debankur1** [comment]
> Ah I see, I believe right now the only way forward here would be to spin up a local registry on that node.

yes but the problem is we can end up with issues related to storage if the server is bare-metal or in a Edge scenario or the storage is limited

**Comment by debankur1** [comment]
I believe to implement the feature I am asking for, might take some time as it will be a shift from the current architecture. Is that the right understanding?

**Comment by 100mik** [comment]
kapp-controller does cache images referred to with shasum references. It won't pull these images twice. 
Does this solve the issue at hand?
